keps/sig-cloud-provider/20200928-extend-konnectivity-for-both-directions.md

```diff
@@ -0,0 +1,194 @@
+                         |            ^
+                         |   Tunnel   |
+                         +------------+
+```
+
+The agent listens for TCP connections at a specific port for each configured destination. When a connection request is received by the Konnectivity Agent the following happens:
+1. A GRPC DIAL_REQ message is sent to the Konnectivity server containing the destination address associated with the current port.
+2. Upon reception of the DIAL_REQ the Konnectivity Server opens a TCP connection with the destination host/port and replies to the Konnectivity Agent with a GRPC DIAL_RES message.
+3. At this point the tunnel is established and data is piped through it, carried over GRPC DATA packets.
+
+### Agent additional flags
+
+* `--target=local_port:dst_host_ip:dst_port`: We can have multiple of those in order to support multiple destinations on the Master Network.
+Dst_host_ip: end target IP (apiserver or something else). In case of IPv6
+the address should be wrapped in squared brackets, consistently with [RFC3986](https://www.ietf.org/rfc/rfc3986.txt) notation.
+e.g. --target=6443:apiserver.svc.cluster.local:6443
+e.g. --target=6443:[2001:db8:1f70::999:de8:7648:6e8]:6443
+
+* `--bind-address=ip`: Local IP address where the Konnectivity Agent will listen for incoming requests. It will be bound to a dummy IP interface with IP x.y.z.v defined by the user. Must be used with the previous one to enable incoming requests. If not, and for backward compatibility, only the traffic initiated from the Control Plane will be allowed.
+
+### Handling the Traffic from the Pods to the Agent
+
+As mentioned above, pods make use of  the Kubernetes default service to reach the KAS. To keep things transparent from a Pod perspective, they will hit the Konnectivity Agent using the Kubernetes default service. The endpoint will be the Konnectivity Agent instead of the KAS. 
+The configuration part of the Kubernetes default service will be done using the Apiserver flag `--advertise-address ip` on the Control Plane side.
+`--advertise-address ip` should match the `--bind-address ip` of the Konnectivity Agent described above. 
+
+### Handling the Traffic from the Kubelet to the Agent
+
+Kubelet does not use the Kubernetes default service to reach the KAS. Instead it relies on a bootstrap kubeconfig file that is used to connect to the KAS. It then generates a proper kubeconfig file that will be using the same URL.
+Instead of specifying the KAS FQDN/address in the bootstrap kubeconfig file, we will be using the local IP address of the Konnectivity agent (`--bind-address ip`).
+
+### Deployment Model
+
+The agent can be run as static pod or systemd units. In any case the agent should be started to give access to the KAS to the kubelet first and to the hosted pods later. This means that using DaemonSets or Deployments is not an option in this setup, because the kubelet would not be able to get the pod manifests from the KAS.
```

Comment:
I'm not sure if running as a static pod will solve the kubelet bootstrapping issue. Kubelet needs to access to KAS to watch for secrets/configmaps, and to register the node. I don't know if kubelet can bootstrap in this order:
1. run the proxy agent as a static pod, waiting for it to establish tunnels to the KAS,
2. start watching for secrets/configmaps/pods and register the node.

Also I'm not sure if other node components like node-problem-detector are ok with this bootstrap order.

I think @cheftako knows this better.

Reply:
Definitely the proxy agent is required to be up and running in order for the kubelet to perform its bootstrap sequence. I was expecting the static pods to be running without the kubelet being registered, but I did not test it and I agree we should double check if this is feasible.